Parallelize nifti labels masker

dPys opened the pull request:
Addresses/closes Issue #1501 

*Revised `img_to_signals_labels` to support n_jobs argument.
*To keep track of node indices and their corresponding extracted time-series vectors with joblib, I modified reduction_function to be nested. To minimize memory footprint further, we could also employ memory mapping if we feel it necessary. 

Cheers,
@dPys

NicolasGensollen reviewed (changes requested):
Thanks for the PR @dPys !
The code looks good, I just have a few comments:

- I think we need to expose `njobs` to the `NiftiLabelsMasker`.
- The default value for `njobs` is usually 1 rather than -1. 
- Also, I would be in favour of adding a couple tests where `njobs` is set to different values and results are checked to be the same.

dPys commented:
> Thanks for the PR @dPys !
> The code looks good, I just have a few comments:
> 
> * I think we need to expose `njobs` to the `NiftiLabelsMasker`.
Done. I also added some error-handling for it.

> * The default value for `njobs` is usually 1 rather than -1.
This was a typo from when I was tinkering, good catch!
 
> * Also, I would be in favour of adding a couple tests where `njobs` is set to different values and results are checked to be the same.
See commit [a88f21c](https://github.com/nilearn/nilearn/pull/2693/commits/a88f21c2deaf15d1c7d3af253d0805fe09aff2ba) where I've added one. I agree that this is a nice test to have although is there any particular mathematical reason why we should expect the results to diverge with a different number of threads in this context? Also, you raise an interesting new precedent because there are no other tests in nilearn, as far as I can tell, that involve manipulating n_jobs in a systematic fashion as I'm doing here. e.g. in [test_searchlight](https://github.com/nilearn/nilearn/blob/master/nilearn/decoding/tests/test_searchlight.py#L32), n_jobs is held constant at 1. My sense is that if we indeed decide to incorporate this test here, parallelized functionality elsewhere should probably be held to the same standard, no?

NicolasGensollen reviewed (changes requested):
Thanks for addressing all the points so quickly! :+1: 
I only have a few small comments (see below).

NicolasGensollen commented on nilearn/input_data/tests/test_nifti_labels_masker.py:
```diff
@@ -368,3 +368,57 @@ def test_nifti_labels_masker_with_mask():
+        assert np.allclose(n_result, expected_result)
+
+    with pytest.raises(ValueError,
+                       match=f"Invalid value for n_jobs '{-2}'. Must be an "
```
I think this f-string is breaking the python 3.5 tests

NicolasGensollen commented on nilearn/input_data/tests/test_nifti_labels_masker.py:
```diff
@@ -368,3 +368,57 @@ def test_nifti_labels_masker_with_mask():
+        )
+
+    with pytest.raises(ValueError,
+                       match=f"Invalid value for n_jobs '{0}'. Must be an "
```
and this one

NicolasGensollen commented on nilearn/input_data/tests/test_nifti_labels_masker.py:
```diff
@@ -368,3 +368,57 @@ def test_nifti_labels_masker_with_mask():
+        )
+
+    with pytest.raises(ValueError,
+                       match=f"Invalid value for n_jobs '{1.5}'. Must be an "
```
and also this one

NicolasGensollen commented:
I thought about the `n_jobs` test to make sure the `n_jobs` parameter is handled correctly across the pipeline. I understand your point that there no reason we should get different results (assuming no bugs in joblib). I still think it's nice to have it (especially now that it's written...), what do you think @bthirion ?

dPys commented:
Requested changes made.

bthirion commented:
I'm a bit unsure whether using 'n_jobs=1' in test results in a waste of resources.  
testing for 'n_jobs=2' is indeed less susceptible of generating waste of cost/CPU time.
@NicolasGensollen bzw, we should keep an eye on build time: it is good to have everything tested, but we have to be careful about computation resources.

dPys commented:
> I'm a bit unsure whether using 'n_jobs=1' in test results in a waste of resources.
> testing for 'n_jobs=2' is indeed less susceptible of generating waste of cost/CPU time.
> @NicolasGensollen bzw, we should keep an eye on build time: it is good to have everything tested, but we have to be careful about computation resources.

One thing to note @bthirion -- n_jobs = 1 is actually not really being "tested" here. It is simply used once initially to simulate an `expected_result` to compare with that being derived from multithreaded computation (i.e. n_jobs=2). 

Added build time is ~1.34 seconds.

bthirion commented:
Thx. LGTM then.

dPys commented:
> Thx. LGTM then.

Sounds good. Excited to see how this scales to 1000+ labels... 

Thanks @NicolasGensollen and @bthirion for your review.

NicolasGensollen reviewed (approved):
I think we can add an entry in whats_new. 
Looks good to me otherwise, thanks! :+1:

jeromedockes commented on nilearn/input_data/nifti_labels_masker.py:
```diff
@@ -178,6 +186,14 @@ def __init__(self, labels_img, background_label=0, mask_img=None,
             raise ValueError("invalid value for 'resampling_target' "
                              "parameter: " + str(resampling_target))
 
+        if float(n_jobs) < -1 or not float(n_jobs).is_integer() or \
```
these checks look like they should either happen for all maskers or none of them

dPys replied:
@jeromedockes -- I've gone ahead and relocated the checks. Please let me know if the most recent push doesn't address your concern here, in which case let me know how you'd prefer this be handled.

jeromedockes commented on nilearn/input_data/nifti_labels_masker.py:
```diff
@@ -286,6 +286,15 @@ def transform_single_imgs(self, imgs, confounds=None):
 
         """
+
+        if float(self.n_jobs) < -1 or not float(self.n_jobs).is_integer() or \
```
IIUC the check is repeated here to avoid doing the resampling if the ExtractionFunctor init is going to fail, is that right?

dPys replied:
Correct!

dPys commented on nilearn/regions/signal_extraction.py:
```diff
@@ -18,7 +18,8 @@
 def img_to_signals_labels(imgs, labels_img, mask_img=None,
-                          background_label=0, order="F", strategy='mean'):
+                          background_label=0, order="F", strategy='mean',
+                          n_jobs=1, verbose=0):
```
Added. I also moved the n_jobs correctness error-handling from `_ExtractorFunctor` to live directly in img_to_signals_labels given that the `__call__` to the latter is the only method of the `_ExtractorFunctor` class
